projects/chaos/src/relay.c

```diff
@@ -0,0 +1,50 @@
+#include "relay.h"
+
+#include <stdbool.h>
+
+#include "can_msg_defs.h"
+#include "event_queue.h"
+#include "relay_fsm.h"
+#include "relay_id.h"
+
+static FSM s_solar_front_relay_fsm;
+static FSM s_solar_rear_relay_fsm;
+static FSM s_battery_relay_fsm;
+static FSM s_main_relay_fsm;
+
+void relay_init(bool loopback) {
+  relay_fsm_init();
+
+  uint32_t solar_front_bitset;
+  uint32_t solar_rear_bitset;
+  uint32_t main_bitset;
+  uint32_t battery_bitset;
+
+  if (loopback) {
+    solar_front_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_CHAOS);
+    solar_rear_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_CHAOS);
+    main_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_CHAOS);
+    battery_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_CHAOS);
+  } else {
+    solar_front_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_SOLAR_MASTER_FRONT);
+    solar_rear_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_SOLAR_MASTER_REAR);
+    main_bitset = CAN_ACK_EXPECTED_DEVICES(CAN_DEVICE_MOTOR_CONTROLLER);
```

Comment:
The intention was that power distribution's power rail would turn the motor controllers on - motor controller interface is only on when the motor controllers are on kind of thing. I suppose we could make them CAN-controlled, but I don't see the point. They're in the same box.

Thinking about it, is there even a need to make the solar relays CAN-controlled? It could just turn on as soon as it gets power.

Reply:
I have serious concerns about automatically opening relays. Not only does this allow events to occur out of order (ie. the main relay closes too early) but what happens if a MOSFET or other component fails closed leaving the board active. This also requires boards to acknowledge a successful closure and they will fail or be retried otherwise. This is a much safer design IMO. Furthermore it allows determinism in when the relays close since the startup and firmware init time of the boards while short isn't necessarily the same for all the boards possibly leading to out of order relay changes.

Reply:
Sure, the idea was the power distribution handles power sequencing, but if you feel more comfortable with this, then I'm fine with it. Note that if a component does fail, we really have no way to detect that.

Reply:
I primarily want it for the sequencing control. I'll handle any implementation issues on the other boards but it should be pretty simple.